### ISSUE TITLE
Pull twice from subrepo loses changes from main repo

### ISSUE THREAD

**Issue by grimmySwe** [ingydotnet/git-subrepo#98]
Using subrepo v0.2.3 (679ad12fd684bd72feff90d445c373d9a3d2aa0f)
git version 1.9.3, 2.6.0
Also tested dzzh fork (f21fb34c676ea9949d870f38b1f934d89772c148)

The following workflow seems to lose things:

I start with this
```
bar
-----------------
foo
foo/bar
```
```cd foo; git add foo/bar/foo1```
```cd bar; git add bar/bar1```

```
bar
bar/bar1
-----------------
foo
foo/bar
foo/bar/foo1
```
```cd foo; git subrepo pull bar```
```
bar
bar/bar1
-----------------
foo
foo/bar
foo/bar/bar1
foo/bar/foo1
```
```cd bar; git add bar/bar2```
```cd foo; git subrepo pull bar```
```
bar
bar/bar1
bar/bar2
-----------------
foo
foo/bar
foo/bar/bar1
foo/bar/bar2
```
So where did foo/bar/foo1 go!!!!

I will see if I can create a test case for this.

**Comment by grimmySwe** [comment]
```
#!/usr/bin/env bash
set -e
source test/setup
use Test::More
clone-foo-and-bar
subrepo-clone-bar-into-foo

(
  cd $OWNER/bar
  add-new-files Bar2
  git push
) &> /dev/null || die

(
  cd $OWNER/foo
  add-new-files bar/Foo2
  git push
  git subrepo pull bar
) &> /dev/null || die

(
  cd $OWNER/bar
  add-new-files Bar3
  git push
) &> /dev/null || die

(
  cd $OWNER/foo
  git subrepo pull bar
) &> /dev/null || die

test-exists \
  "$OWNER/foo/bar/Bar2" \
  "$OWNER/foo/bar/Bar3" \
  "$OWNER/foo/bar/Foo2" \

done_testing
teardown
```
This fails with
```
$ prove test/subrepo-pull-twice.t
test/subrepo-pull-twice.t .. 1/? #   Failed test 'File 'tmp/owner/foo/bar/Foo2' exists'
#   at test/setup line 81.
test/subrepo-pull-twice.t .. Failed 1/3 subtests 

Test Summary Report
-------------------
test/subrepo-pull-twice.t (Wstat: 0 Tests: 3 Failed: 1)
  Failed test:  3
Files=1, Tests=3,  1 wallclock secs ( 0.02 usr  0.00 sys +  0.17 cusr  0.50 csys =  0.69 CPU)
Result: FAIL
```

**Comment by grimmySwe** [comment]
It seems that during the pull command we decide if there has been any actions on the main repo:
```
 o "Make sure there is at least one commit after the last pull."
  # (That translates to 2 commits after the parent of last pull):
  local count="$(git rev-list "$subrepo_parent"..HEAD -2 | grep -c $'\n')"
  if [ "$count" != 2 ]; then
    OK=false; CODE=-2; return
  fi
```
So when you run the pull you get something like this
```
foo$ git subrepo pull bar --debug
>>> git rev-parse --verify HEAD
>>> git fetch --quiet ../bar/.git master
>>> git rev-parse FETCH_HEAD
>>> git branch -D subrepo/bar
>>> git branch subrepo/bar refs/subrepo/bar/fetch     # We create branch at the fetched ref
>>> rm -fr bar1 bar2 foo1 .gitrepo                               # Remove all local files
>>> git reset --hard subrepo/bar                                 # Reset to the branch (That not include foo1!!!)
....
```
So for pull to work "correctly" it will need to have a commit after the last pull. But what if you don't want to push you local changes? I can see a use case where you work for a while only pulling in changes into the subrepo just commiting your changes locally to the main repo. But pulling twice without pushing to the subrepo will make files vanish, or am I missing something here?

It seems dangerous just ignore the main repo if no commits has been done since last pull, as you don't always do the push and as seen in #69 you can run ```git subrepo push bar``` missing the extra branch parameter needed and get a note that you are up to date.

**Comment by grimmySwe** [comment]
I have looked into this some more and have actually come with a solution to this problem, although it's very untested for other cases...

During the branch process you try to find new commits to merge into the subrepo
```
>>> git filter-branch --force --parent-filter <DATA>   # Remove all before parent (1)
>>> git filter-branch -f --subdirectory-filter bar HEAD   # Convert to subdir (2)
>>> git filter-branch --force --parent-filter <DATA>   # Remove parent (3)
>>> git filter-branch -f --tree-filter rm -f .gitrepo 
```
What if you skip step 1 and 3, that way you will always get a branch that contains everything that has been done to the local subrepo. Not only since last pull, instead you always get the complete story. 

I understand that there is probably large holes in this theory but I can't really find another way to make sure that you don't forget/lose things. I will try and see what the pushes will end up with in these cases. But as long as a --subdirectory-filter commit transforms into the same hash every time it should be able to work properly against the subrepo.

**Comment by grimmySwe** [comment]
Ok, so I thought a little more and found the hole... Current design takes all the commits in main repo that regards the subrepo part and removes everything before the last pull + the last pull, why? Because  you don't want the pull commits to end up in the subrepo...

So the question is if there is a way to filter out all the pull commits.

There seems to be way to find all commits that changes a certain file and in our case we have a .gitrepo that is exactly that. A file that only changes during the pull operations.
```git log --follow filename```

So if it's possible:
1. Convert to subdir
2. Filter out the commits changing .gitrepo
3. Remove .gitrepo

**Comment by grimmySwe** [comment]
So here is an embryo to some code to perform the wanted action:
```
#!/bin/bash

hashes=$(git log --pretty=format:%H bar/.gitrepo);
echo "$hashes";

git filter-branch -f --commit-filter "
        git_commit=\$GIT_COMMIT
        local_hashes=$hashes
	if [ \${local_hashes/\$git_commit} = \$local_hashes ] ;
	then
                >&2 echo Keep: \$@;
        	git commit-tree \$@;
	else
                >&2 echo Skip: \$@;
		skip_commit \$@;
	fi" HEAD
```

**Comment by grimmySwe** [comment]
@ingydotnet @AndrewSmart @dzzh @jrosdahl
Hi all,  I want to start with saying that I really like subrepo and it feels like the natural way of handling external repos. But... there is some shortcomings that currently makes it a little dangerous to use. As this is an open source project I feel that I would like to contribute with some ideas but I feel that I could use some experienced opinions on how you might solve these shortcomings.

In this issue I have discovered problems that will cause changes to vanish if you don't follow a strict workflow. And even then it might be a little tricky. I have solved some of these cases by letting the subrepo branch command include the entire history except pull commits. But then I realized that a pull commit can actually be two animals, one is the regular pull commit that only updates but it can also be a merge pull commit that first updates and then merges your local changes.

I know that the main idea with subrepo is that it should be as transparent as possible so it doesn't add a lot of extra stuff on main or sub repos. But I was considering an idea of introducing strict pull commits that always contain only the actual fetch from the subrepo, letting the merge be a separate commit. That way you can always ignore the "pull" commits when you then push data back into the subrepo. 
![git subrepo pull-push suggestion](https://cloud.githubusercontent.com/assets/6480514/10998303/82090cf4-8494-11e5-83ce-bbb1554edd8c.png)

So the workflow when you pull data is:
1. Is my local subrepo different then then previous pulled version?
If no, then I can simply update my local subrepo (A) with the new subrepo (B) (this would be running pull on (2)
If yes,
2. Create a new local commit (4) that is based on my local HEAD + the exact copy of the new subrepo (B)
3. Merge the subrepo part of this new commit into 5 (3' & B => 5')

When I push in this case:
1. Create remote version of the subrepo
2. Create local version of the subrepo by subdirectory-filter
3. Push the difference to the remote (if it fails, you might want to force a pull first)

Benefits (if it works)
* would be that you would not lose things easily
* you will get the complete tree on the subrepo side as well.

Drawbacks
* It will complicate the history, when you merge in things from subrepo you will get two commits instead of only one.

But as I have only looked at this git-subrepo for a short while it would be interesting to get some opinions if this is good way ahead.

**Comment by dzzh** [comment]
You're right, it seems that doing several pulls without pushes will discard the changes made between them. I don't use this flow myself, so I didn't notice this issue. It looks like it might help if in my fork we remove previous_pull_commit (either from everywhere or from nearly all places) and use previous_push_parent instead. This is especially important during `subrepo:branch`. If we do so, then a cherry-pick commit applied on top of `upstream_head` will include all the changes you made no matter how many subrepo pulls were there before them. I'll try to find time in the weekend to change the code of my fork.

**Comment by grimmySwe** [comment]
I created a fork (https://github.com/grimmySwe/git-subrepo) and created an issue98 branch where I have commited my current attempts.

@dzzh I copied your test cases for issue95 and issue96 and tested them with my solution and they seem to work.

Current status for me is that I can't handle merges in a good way. I will look into your proposed solution and see if I can understand what you are trying to do!

**Comment by grimmySwe** [comment]
@dzzh I read through your code and I think I understand what you are trying to do. I am not sure though that I want squashing commits in that direction. Then it will look different in the subrepo depending on if I have edited it in the actual subrepo or in a another repo.

**Comment by grimmySwe** [comment]
I have continued on my investigation and unfortunately the path was blocked... New obstacles came up with my solution. So I am back to square two... with no solution to the merge problem.

**Comment by AndrewSmart** [comment]
@grimmySwe I know exactly what you mean about new obstacles coming up, and having to go back to square one :-)
I tried your test/subrepo-pull-twice.t, you're right about Foo2. This is a severe bug that can lead to lost files.

**Workaround**: Always push before pulling:
```bash
git subrepo push X
git subrepo pull X
```

**Comment by grimmySwe** [comment]
@AndrewSmart @dzzh @ingydotnet 
I have actually come up with a solution that I think works...

On my fork I have implemented the following solution:
1. subrepo branch will always contain all the commits.
2. When performing pull/push I actually compare tree hashes connected to commit hashes to find common commits. Then I perform a rebase --onto between the subrepo/bar/fetch and subrepo/bar. This way I will get a common ancestor! After that I perform the "regular" rebase, in the push case I use the "-X theirs" parameter. Theirs in this case allows us to use our own changes without getting conflicts! This should not be used unless you have actually performed a pull first. So I have also added a check that to do a push you need to be up to date to repo, if not, you must use the --force flag.

This way I make sure that I don't forget changes on either side.

I have not seen any huge obstacles yet but I am currently updating test cases to reflect the current logic.

Known issues...
- Currently if you perform a merge in the actual pull commit, causing it to create a new treehash. The commit message when you push will contain "git subrepo commit bar ... ... ... ... "

**Comment by AndrewSmart** [comment]
@grimmySwe 
> 2. When performing pull/push I actually compare tree hashes connected to commit hashes to find common commits. Then I perform a rebase --onto between the subrepo/bar/fetch and subrepo/bar. This way I will get a common ancestor!

Sounds exactly like an approach I used about a year ago. I approve! I went down this route with an alpha version of git-subrepo, as a rebase was being done with the .gitrepo on one branch altered all the commit hashes so that the rebase selected a non-optimal common ancestor. I created new hashes without the .gitrepo file so that the hashes between the branches would line up, to find the optimal common ancestor. This solved some issues before Ingy made major improvements.
> After that I perform the "regular" rebase, in the push case I use the "-X theirs" parameter. Theirs in this case allows us to use our own changes without getting conflicts!

Sounds like a great approach to me. Hope with your approach things work out!

**Comment by grimmySwe** [comment]
@AndrewSmart @dzzh @ingydotnet 
I have pushed some code to my fork, branch issue98. It seems to work :-) I will see if I can test it and write even more test cases for it. If you have any test case please let me know! 

I am an experienced developer but a n00b bash:er and haven't used git for so long so any improvement suggestions are welcome!

**Comment by wkornewald** [comment]
@grimmySwe This goes into the direction I'm experimenting with. Is there a reason why you use -X theirs instead of --merge --interactive? After all, if there's a conflict it must be resolved manually and won't it be more difficult to fix after the rebase -X theirs?

Also, I'm still trying to find a good solution for sharing the rebased state with others. After all, it might contain non-trivial changes which we don't want every developer to reconstruct. Not sure if it's a good idea to commit all rebased patches into the project, so others can reuse those (subrepo could automatically load those patches when building the branch).

Finally, one issue I've stumbled upon is that sometimes the common ancestor was imported with slight modifications, so the treehash won't match. It should still be possible to build the branch in that case, e.g. by rebasing from the first commit or even by allowing to specify the common ancestor revision.

**Comment by grimmySwe** [comment]
@wkornewald The reason I am using `rebase -X theirs` is that I consider the pull step to be the one where you solve conflicts and there I don't have any -X at all. At the push step there is a prerequisite that you have pulled in the latest version from the subrepo. And then I consider the push step to be more of a publish step, and with the -X theirs I don't have to solve conflicts, I just apply my changes as is. But please remember that I haven't worked with git a lot and there might be flaws in my way of working.

> Also, I'm still trying to find a good solution for sharing the rebased state with others. After all, it might contain non-trivial changes which we don't want every developer to reconstruct. Not sure if it's a good idea to commit all rebased patches into the project, so others can reuse those (subrepo could automatically load those patches when building the branch).

Regarding your second paragraph of sharing the rebased state, could you provide some examples or test cases on that? I am not sure I understand what you mean.

> Finally, one issue I've stumbled upon is that sometimes the common ancestor was imported with slight modifications, so the treehash won't match. It should still be possible to build the branch in that case, e.g. by rebasing from the first commit or even by allowing to specify the common ancestor revision.

I haven't found any of these problems yet... But as you said the code needs to handle them in a good way. Do you have any specific cases where the tree hash differ? I do see that there could be a rebase that actually removes any common ground so in that case subrepo should warn and force the user to handle the conflict in a good way, maybe reclone or something like that.

Finally I was thinking of adding the subrepo-merge-base as a possible top level command so that you can verify it's functionality. Then you could do something like this

```
git subrepo fetch bar
git subrepo branch bar
git subrepo merge-base subrepo/bar/fetch subrepo/bar
Found equal tree has:
subrepo/bar/fetch: <commit_hash>
subrepo/bar: <commit_hash>
```

**Comment by grimmySwe** [comment]
I did an update adding `git subrepo merge-base <branch1> <branch2>` command, also added an error if the subrepo is totally out of common ancestors, due to a `git push --force` or other evil operation.
Updated the docs and created a pull request for the main repo.

I will now start to use this in production and see if it works...

Please let me know if you have and problems or test case suggestions!

**Comment by ingydotnet** [comment]
@grimmySwe et al,

Apologies for being away, and kudos for the hard work and thoughtful research. I will integrate your fork into origin/issue/98 on this repo and start thinking on it. Please keep me up to date on progress from your end.

Cheers!

**Comment by ingydotnet** [comment]
I reviewed the code for obvious stuff and ran the tests. Overall impression? A+ :)

The code looks great. I pushed 4 commits to issue/98. Read the messages. Tells how to do the generated files properly.

Let's wait for a week of stability and testing, then we can probably release it.

Cheers!

PS I added @grimmySwe as a collaborator. You can push new stuff directly to the issue/98 branch if you'd like to.

**Comment by grimmySwe** [comment]
@ingydotnet 

> I reviewed the code for obvious stuff and ran the tests. Overall impression? A+ :) The code looks great. 

Thanks!

> I pushed 4 commits to issue/98. Read the messages. Tells how to do the generated files properly.

Read and understood. Good to know.

> Let's wait for a week of stability and testing, then we can probably release it.

I try to use it as much as I can during daily business. I hope that @dzzh will test it for reported issues #89, #91, #95, #96 and see if it solves them (the tests seems to pass).

**Comment by ingydotnet** [comment]
@dzzh @grimmySwe are there git-subrepo/test/ tests for #89, #91, #95, #96?

**Comment by grimmySwe** [comment]
I copied some test cases from @dzzh fork for issue 95 and 96, they should
be present on the issue/98 but I don't know if anything happen after that.

onsdag 18 november 2015 skrev Ingy döt Net <notifications@github.com>:

**Comment by ingydotnet** [comment]
Pulled PR 85 into the issue/98 branch (and into master). Force pushed issue/98
branch.

**Comment by ingydotnet** [comment]
@grimmySwe, @dzzh, et al. How are we feeling about this?

I'm ready to merge to master if there is no concern.

**Comment by grimmySwe** [comment]
@ingydotnet Go! I have used this branch and with the latest fix from @jrosdahl I had no issue.

**Comment by ingydotnet** [comment]
@grimmySwe sweet. I'll push out the release this afternoon. Great work!

Please continue to review the issues as your time permits.

**Comment by ingydotnet** [comment]
@grimmySwe, I just pushed some light changes for a 0.3.0 release tomorrow. Have
a look/review/test.

If all looks good, I'll push the button. If not, feel free to push corrections
to the issue/98 branch.

**Comment by jrosdahl** [comment]
I've also used the issue/98 branch for a while and I've had no issues. Well done!

**Comment by grimmySwe** [comment]
@ingydotnet  Looks good to me. Your cleanup made it top notch.

**Comment by ingydotnet** [comment]
Version 0.3.0 Released!

Thanks to everyone involved!

@grimmySwe, please close this issue and @grimmySwe @dzzh @jrosdahl @AndrewSmart
@ml31415 please close any other issues that can be closed with this release.

I discovered an anomoly that I reported as #106. I decided it didn't block the
release. Could people recently in this code try to figure it out?

I also lumped on the first stage of #105. Everyone switch to `.rc` please!

I think this thing is getting critical mass. I'd like to keep the energy up
while we have it. I'll look over the current bugs in a few days, after you guys
close a few, and make a roadmap for 2016.

Cheers, Ingy

**Comment by grimmySwe** [comment]
This issue is solved in Version 0.3.0.

I'll check out #106 and see if there are other bugs to be solved. 

@ingydotnet Do you have any labels for issues? It might be good to have some improvement label to see if things are simply improvement suggestions and not bugreports.
